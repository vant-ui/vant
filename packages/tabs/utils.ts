import { raf } from '../utils/dom/raf';

export function scrollLeftTo(el: HTMLElement, to: number, duration: number) {
  let count = 0;
  const from = el.scrollLeft;
  const frames = Math.round((duration * 1000) / 16);

  function animate() {
    el.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      raf(animate);
    }
  }

  animate();
}
