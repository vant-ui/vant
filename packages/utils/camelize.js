const camelizeRE = /-(\w)/g;
export default function(str) {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '');
}
