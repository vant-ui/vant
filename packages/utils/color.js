export const RED = '#f44';
export const BLUE = '#1989fa';
export const GREEN = '#4b0';
export const GRAY_DARK = '#969799';
