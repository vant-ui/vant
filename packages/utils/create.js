/**
 * Create a component with common options
 */
import '../locale';
import Icon from '../icon';
import i18n from '../mixins/i18n';
import install from './install';
import Loading from '../loading';

export default function(sfc) {
  sfc.install = sfc.install || install;
  sfc.mixins = sfc.mixins || [];
  sfc.mixins.push(i18n);
  sfc.components = sfc.components || {};
  sfc.components.icon = Icon;
  sfc.components.loading = Loading;

  return sfc;
};
