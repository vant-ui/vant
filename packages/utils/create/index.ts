import { createBEM } from './bem';
import { createComponent } from './component';
import { createI18N } from './i18n';

type CreateNamespaceReturn = [
  ReturnType<typeof createComponent>,
  ReturnType<typeof createBEM>,
  ReturnType<typeof createI18N>
];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = 'van-' + name;
  return [createComponent(name), createBEM(name), createI18N(name)];
}
