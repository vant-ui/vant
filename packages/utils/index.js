import Vue from 'vue';

export { use, useSlots } from './use';

export const isServer = Vue.prototype.$isServer;

export function noop() {}

export function isDef(value) {
  return value !== undefined && value !== null;
}

export function isObj(x) {
  const type = typeof x;
  return x !== null && (type === 'object' || type === 'function');
}

export function get(object, path) {
  const keys = path.split('.');
  let result = object;

  keys.forEach(key => {
    result = isDef(result[key]) ? result[key] : '';
  });

  return result;
}

const camelizeRE = /-(\w)/g;
export function camelize(str) {
  return str.replace(camelizeRE, (_, c) => c.toUpperCase());
}

export function isAndroid() {
  /* istanbul ignore next */
  return isServer ? false : /android/.test(navigator.userAgent.toLowerCase());
}

export function range(num, min, max) {
  return Math.min(Math.max(num, min), max);
}
