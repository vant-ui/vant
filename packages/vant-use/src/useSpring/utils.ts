import { SpringConfig } from './config';

export const now =
  typeof performance !== 'undefined'
    ? performance.now.bind(performance)
    : Date.now.bind(Date);

export const isArray = Array.isArray.bind(Array);

const reusedTuple: [number, number] = [0, 0];
export function stepper(
  secondPerFrame: number,
  x: number,
  v: number,
  destX: number,
  spring: SpringConfig,
): [number, number] {
  // Spring stiffness, in kg / s^2

  // for animations, destX is really spring length (spring at rest). initial
  // position is considered as the stretched/compressed position of a spring
  const Fspring = -spring.tension * (x - destX);

  // Damping, in kg / s
  const Fdamper = -spring.friction * v;

  const a = (Fspring + Fdamper) / (spring.mass || 1);

  const newV = v + a * secondPerFrame;
  const newX = x + newV * secondPerFrame;

  const precision = spring.precision || 0.01;

  if (Math.abs(newV) < precision && Math.abs(newX - destX) < precision) {
    reusedTuple[0] = destX;
    reusedTuple[1] = 0;
    return reusedTuple;
  }

  reusedTuple[0] = newX;
  reusedTuple[1] = newV;

  return reusedTuple;
}
