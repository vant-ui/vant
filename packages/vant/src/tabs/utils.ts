import { raf, cancelRaf } from '@vant/use';
import { ScrollElement, getScrollTop, setScrollTop } from '../utils';

let scrollLeftToRafId: number;
let scrollTopToRafId: number;

export function scrollLeftTo(
  scroller: HTMLElement,
  to: number,
  duration: number
) {
  cancelRaf(scrollLeftToRafId);

  let count = 0;
  const from = scroller.scrollLeft;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);

  function animate() {
    scroller.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      scrollLeftToRafId = raf(animate);
    }
  }

  animate();
}

export function scrollTopTo(
  scroller: ScrollElement,
  to: number,
  duration: number,
  callback: () => void
) {
  cancelRaf(scrollTopToRafId);

  let current = getScrollTop(scroller);
  const isDown = current < to;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);
  const step = (to - current) / frames;

  function animate() {
    current += step;

    if ((isDown && current > to) || (!isDown && current < to)) {
      current = to;
    }

    setScrollTop(scroller, current);

    if ((isDown && current < to) || (!isDown && current > to)) {
      scrollTopToRafId = raf(animate);
    } else if (callback) {
      scrollTopToRafId = raf(callback as FrameRequestCallback);
    }
  }

  animate();
}
