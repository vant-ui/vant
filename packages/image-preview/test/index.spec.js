import Vue from 'vue';
import ImagePreview from '..';
import ImagePreviewVue from '../image-preview';
import { mount } from '@vue/test-utils';
import { triggerDrag } from '../../../test/utils';

const images = [
  'https://img.yzcdn.cn/upload_files/2017/03/15/FkubrzN7AgGwLlTeb1E89-T_ZjBg.png',
  'https://img.yzcdn.cn/upload_files/2017/03/14/FmTPs0SeyQaAOSK1rRe1sL8RcwSY.jpeg',
  'https://img.yzcdn.cn/upload_files/2017/03/15/FvexrWlG_WxtCE9Omo5l27n_mAG_.jpeg'
];

test('render image', () => {
  const wrapper = mount(ImagePreviewVue, {
    propsData: {
      images
    }
  });

  expect(wrapper.html()).toMatchSnapshot();

  const swipe = wrapper.find('.van-swipe__track');
  triggerDrag(swipe, 500, 0);
  expect(wrapper.emitted('input')).toBeFalsy();
  triggerDrag(swipe, 0, 0);
  expect(wrapper.emitted('input')[0][0]).toBeFalsy();
});

test('function call', (done) => {
  ImagePreview(images);
  ImagePreview(images.slice(0, 1));
  Vue.nextTick(() => {
    const wrapper = document.querySelector('.van-image-preview');
    const swipe = wrapper.querySelector('.van-swipe__track');
    triggerDrag(swipe, 0, 0);

    expect(wrapper.querySelectorAll('img').length).toEqual(1);

    Vue.nextTick(() => {
      expect(wrapper.style.display).toEqual('none');
      done();
    });
  });
});

test('register component', () => {
  Vue.use(ImagePreview);
  expect(Vue.component('van-image-preview')).toBeTruthy();
});
