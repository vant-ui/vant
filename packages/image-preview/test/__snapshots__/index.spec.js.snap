// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render image 1`] = `
<div class="van-image-preview" style="display: none;">
  <div class="van-swipe">
    <div class="van-swipe__track" style="width: 0px;">
      <div class="van-swipe-item" style="width: 0px; height: 100%;">
        <img src="https://img.yzcdn.cn/upload_files/2017/03/15/FkubrzN7AgGwLlTeb1E89-T_ZjBg.png" class="van-image-preview__image">
      </div>
      <div class="van-swipe-item" style="width: 0px; height: 100%;">
        <img src="https://img.yzcdn.cn/upload_files/2017/03/14/FmTPs0SeyQaAOSK1rRe1sL8RcwSY.jpeg" class="van-image-preview__image">
      </div>
      <div class="van-swipe-item" style="width: 0px; height: 100%;">
        <img src="https://img.yzcdn.cn/upload_files/2017/03/15/FvexrWlG_WxtCE9Omo5l27n_mAG_.jpeg" class="van-image-preview__image">
      </div>
    </div>
    <!---->
  </div>
</div>
`;
